src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -337,4 +348,72 @@ void KafkaSource::calculateColumnPositions()
         physical_header.insert({physical_column.type->createColumn(), physical_column.type, physical_column.name});
     }
 }
+
+
+/// It basically initiate a checkpoint
+/// Since the checkpoint method is called in a different thread (CheckpointCoordinator)
+/// We nee make sure it is thread safe
+void KafkaSource::checkpoint(CheckpointContextPtr ckpt_ctx_)
+{
+    /// We assume the previous ckpt is already done
+    /// Use std::atomic<std::shared_ptr<CheckpointContext>>
+    assert(!ckpt_ctx.load(std::memory_order_relaxed));
+    ckpt_ctx = new CheckpointContext(*ckpt_ctx_);
+}
+
+/// 1) Generate a checkpoint barrier
+/// 2) Checkpoint the sequence number just before the barrier
+Chunk KafkaSource::doCheckpoint(CheckpointContextPtr ckpt_ctx_)
+{
+    /// Prepare checkpoint barrier chunk
+    auto result = header_chunk.clone();
+    auto chunk_ctx = std::make_shared<ChunkContext>();
+    chunk_ctx->setCheckpointContext(ckpt_ctx_);
+    result.setChunkContext(std::move(chunk_ctx));
+
+    ckpt_ctx_->coordinator->checkpoint(State::VERSION, getLogicID(), ckpt_ctx_, [&](WriteBuffer & wb) { ckpt_data.serialize(wb); });
+
+    /// FIXME, if commit failed ?
+    /// Propagate checkpoint barriers
+    return result;
+}
+
+void KafkaSource::recover(CheckpointContextPtr ckpt_ctx_)
+{
+    ckpt_ctx_->coordinator->recover(
+        getLogicID(), ckpt_ctx_, [&](VersionType version, ReadBuffer & rb) { ckpt_data.deserialize(version, rb); });
+
+    LOG_INFO(log, "Recovered last_sn={}", ckpt_data.last_sn);
```

Comment:
What if drop kafka topic and recreate it before recover?

Reply:
it's a good question, i will create an issue to verify and resolve it.

Reply:
This is external data source problem. The best try is detect this and report error (latest Kafka topic has a unique id internally). Otherwise, we can't control it.